serpentTools/parsers/branching.py

```diff
@@ -15,6 +15,164 @@ class BranchingReader(XSReader):
+            # init list to hold branch names if it is still None
+            if layers is None:
+                layers = [[] for i in range(int(l[0]))]
+            for i, branch in enumerate(l[1:]):
+                # append if not seen before
+                if branch not in layers[i]:
+                    layers[i].append(branch)
+            theseBranches = l[1:]
+
+            next(f)
+            l = next(f).split()
+            BUi = int(l[1])  # index of BU step
+            if BUi not in coe.keys():
+                coe[BUi] = {}
+
+            # ok, should be at a totally new set of data now,
+            # so, let's gather all params.
+            for iUni in range(nUni):
+
+                l = next(f).split()
+                uni = l[0]
+                if uni not in coe[BUi].keys():
+                    coe[BUi][uni] = {}
+                nPara = int(l[1])  # num of params to follow
+
+                thisCoe = coe[BUi][uni]
+                # delve through the branches, reassigning the thisCoe reference
+                # at each successive level
+                for i, branch in enumerate(theseBranches):
+                    if branch not in thisCoe.keys():
+                        thisCoe[branch] = {}
+                    if i == len(theseBranches)-1:
+                        break
+                    thisCoe = thisCoe[branch]
+
+                names = []
+                values = []  # should all be numeric in nature
+                for i in range(nPara):
+                    # build from lists to avoid frequently rehashing dicts
+                    l = next(f).split()
+                    name = l[0]
+                    # check if B1
+                    # also, yeah 'and' could be used, but bool checking
+                    # is faster than string comparison, so that's why to nest
+                    if not grabB1:
+                        if name[0:2] == 'B1':
+                            continue
+                    value = l[2:]
+                    names.append(name)
+                    if len(values) == 1:
+                        values.append(float(value[0]))
+                    else:
+                        values.append([float(item) for item in value])
+
+                thisCoe[branch] = dict(zip(names, values))
+
+        if not fromFh:
+            # if file handle wasnt passed, close what this function opened
+            f.close()
+
+        return coe
+
+    def print_coe(self, coedict, lvl=0, short=True):
+        """ Pretty prints a dictionary returned
+        by parse_coe. Use this to visualize its
+        nested structure.
+
+        Parameters
+        ----------
+        coedict : dict
+            particularly, one returned by parse_coe,
+            although it will work on any nested dict.
+        short : bool
+            short print, only prints first few params
+            found. Set to false for a full STDOUT.
+        lvl : integer
+            used for recursion. don't touch.
```

Comment:
Might want to hide the recursive call inside another function then if you have an argument that shouldn't be called. Could be tricky but having an argument implies it could be modified upon calling.

Reply:
Thanks Drew, I'll be sure to improve this.